### PR TITLE
Do not assert in user code

### LINKED ISSUE #33: Assert false raised at line 241 in solveBounds.ml
TL;DR : maximizing on some environments leads to a critical failure.

The following script tries to maximize the polynomial 100 \* v1 + 100 \* v2 with v1 >= 0, v2 >= 0 and 2 \* v1 + 2 \* v2 <= 3, which is possible on Z (v1 = 1, v2 = 0) . But instead of returning something, it leads to an `assert false`: 

```
module Var = struct
  include String
  let is_int _r = true
  let print = Format.pp_print_string
end

module Z = struct
  include Z
  let m_one = Z.sub Z.zero Z.one
  let is_zero = Z.equal Z.zero
  let is_one = Z.equal Z.one
  let is_m_one = Z.equal m_one
  let sign x = Z.compare x Z.zero
  let mult = Z.mul
  let is_int _ = true
  let minus = Z.neg
  let floor = Fun.id
  let ceiling = Fun.id
  let print = Z.pp_print
end

module Explanation = struct
  type t = unit
  let empty = ()
  let union _ _ = ()
  let print _ _ = ()
end

module Sim = OcplibSimplex.Basic.Make(Var)(Z)(Explanation)

let bound_of_z (z : Z.t) : Sim.Core.bound =
  {bvalue = Sim.Core.R2.of_r z; explanation = ()}

let zero_bound = bound_of_z Z.zero

let var ?min ?max name core = fst @@ Sim.Assert.var core ?min ?max name

let poly ?min ?max poly name core = fst @@ Sim.Assert.poly core poly ?min ?max name

let maximize p core = Sim.Solve.maximize core p

let trigger_bug () =
  let core = Sim.Core.empty ~is_int:true ~check_invs:false in
  core
  |> var ~min:zero_bound "v1"
  |> var ~min:zero_bound "v2"
  |> poly
    ~max:(bound_of_z @@ Z.of_int 3)
    (Sim.Core.P.from_list [
        "v1", Z.of_int 2;
        "v2", Z.of_int 2;
      ])
    "poly"
  |> maximize (Sim.Core.P.from_list [
      "v1", Z.of_int 100;
      "v2", Z.of_int 100;
    ]
    )

let () = ignore @@ trigger_bug ()
```

### PR DESCRIPTION
**Pull request by bclement-ocp**
Instead of triggering an [assert false], the [poly] function now raises [Invalid_argument] when provided with a polynomial that has only one variable.

Fixes #33

### CODE REVIEW

File: src/assertBounds.ml [hunk 1]
```diff
@@ -219,6 +219,16 @@ module Make(Core : CoreSig.S) : S with module Core = Core = struct
...
+         here because we introduce the equality [poly = slk] and then we set
+         the bounds for [slk] to [min] and [max].
+
+         If there is a single variable (i.e. [poly = k * x] for some
+         coefficient [k] and variable [x]), we do not want to add the useless
+         slack variable [slk = k * x]. *)
+      if not (P.is_polynomial p) then
+        invalid_arg
+          "poly: must have two variables or more, use var instead";
       assert (P.is_polynomial p);
```
Comment (Halbaroth) [inline src/assertBounds.ml]: We can remove this line as we check if `p` is a polynomial above.

**Review by Halbaroth — CHANGES_REQUESTED** [review]

File: src/assertBounds.ml (hunk 1 above)
Comment (bclement-ocp) [inline src/assertBounds.ml]:
```suggestion
```

Comment (bclement-ocp) [reply src/assertBounds.ml]: Duh, I thought I did. Done.

Comment (Halbaroth) [reply src/assertBounds.ml]: I understand the feeling, it's really frustrating

**Review by Halbaroth — APPROVED** [review]

**Pull request merged**